knebind/solver/solver.go

```diff
@@ -89,15 +88,65 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
 		dstDev := id2Dev[dstParts[0]]
 		srcPort := s.dev2Ports[srcDev][srcParts[1]]
 		dstPort := s.dev2Ports[dstDev][dstParts[1]]
-		dLink := &devLink{srcDev: srcDev, srcPort: srcPort, dstDev: dstDev, dstPort: dstPort}
-		s.dev2Links[srcDev][dstDev] = append(s.dev2Links[srcDev][dstDev], dLink)
-		s.dev2Links[dstDev][srcDev] = append(s.dev2Links[dstDev][srcDev], dLink)
+		group := srcPort.GetGroup()
+		if group == "" {
+			group = dstPort.GetGroup()
+		}
+		devLinkPtr := &devLink{srcDev: srcDev, dstDev: dstDev, srcPort: srcPort, dstPort: dstPort, group: group}
+		if _, ok := s.dev2Links[srcDev][dstDev]; !ok {
+			pgPtr := &devPG{groupMap: make(map[string][]*devLink), sortedNames: []string{}, totalLinks: 0}
+			s.dev2Links[srcDev][dstDev] = pgPtr
+			s.dev2Links[dstDev][srcDev] = pgPtr
+		}
+		s.dev2Links[srcDev][dstDev].groupMap[group] = append(s.dev2Links[srcDev][dstDev].groupMap[group], devLinkPtr)
+	}
+	// For dev / node port group arrange the group names in descending order of membership size.
+	// So for multiple port groups between two devices (and nodes), we assign interfaces for largest port group first
+	// and go down that list; all un-assigned interfaces can be allocated to regular port.
+	// pgMap - map of port group name and membership size
+	// returns list of sorted port group names in the order of membership size, and number of total links
+	sortGroupBySize := func(pgMap map[string]int) ([]string, int) {
```

Comment:
Do we have a test or benchmark that demonstrates the benefit of sorting? If not, postpone until there is one.

Reply:
Yes, I will add one. Without sorting, it may not 'solve' for some correct configurations.